pallets/liquidity-pools-gateway/src/tests.rs

```diff
@@ -42,812 +50,4349 @@ mod utils {
+												),
+											),
+										],
+									},
+								),
+								(
+									vec![
+										(ROUTER_ID_3, Message::Proof(MESSAGE_PROOF)),
+										(ROUTER_ID_2, Message::Proof(MESSAGE_PROOF)),
+										(ROUTER_ID_3, Message::Proof(MESSAGE_PROOF)),
+									],
+									ExpectedTestResult {
+										message_submitted_times: 0,
+										expected_storage_entries: vec![
+											(ROUTER_ID_1, None),
+											(
+												ROUTER_ID_2,
+												Some(
+													ProofEntry {
+														session_id: TEST_SESSION_ID,
+														current_count: 1,
+													}
+													.into(),
+												),
+											),
+											(
+												ROUTER_ID_3,
+												Some(
+													ProofEntry {
+														session_id: TEST_SESSION_ID,
+														current_count: 2,
+													}
+													.into(),
+												),
+											),
+										],
+									},
+								),
+								(
+									vec![
+										(ROUTER_ID_2, Message::Proof(MESSAGE_PROOF)),
+										(ROUTER_ID_3, Message::Proof(MESSAGE_PROOF)),
+										(ROUTER_ID_2, Message::Proof(MESSAGE_PROOF)),
+									],
+									ExpectedTestResult {
+										message_submitted_times: 0,
+										expected_storage_entries: vec![
+											(ROUTER_ID_1, None),
+											(
+												ROUTER_ID_2,
+												Some(
+													ProofEntry {
+														session_id: TEST_SESSION_ID,
+														current_count: 2,
+													}
+													.into(),
+												),
+											),
+											(
+												ROUTER_ID_3,
+												Some(
+													ProofEntry {
+														session_id: TEST_SESSION_ID,
+														current_count: 1,
+													}
+													.into(),
+												),
+											),
+										],
+									},
+								),
+								(
+									vec![
+										(ROUTER_ID_2, Message::Proof(MESSAGE_PROOF)),
+										(ROUTER_ID_3, Message::Proof(MESSAGE_PROOF)),
+										(ROUTER_ID_3, Message::Proof(MESSAGE_PROOF)),
+									],
+									ExpectedTestResult {
+										message_submitted_times: 0,
+										expected_storage_entries: vec![
+											(ROUTER_ID_1, None),
+											(
+												ROUTER_ID_2,
+												Some(
+													ProofEntry {
+														session_id: TEST_SESSION_ID,
+														current_count: 1,
+													}
+													.into(),
+												),
+											),
+											(
+												ROUTER_ID_3,
+												Some(
+													ProofEntry {
+														session_id: TEST_SESSION_ID,
+														current_count: 2,
+													}
+													.into(),
+												),
+											),
+										],
+									},
+								),
+							]);
+
+						let suite = generate_test_suite(
+							vec![ROUTER_ID_1, ROUTER_ID_2, ROUTER_ID_3],
+							TEST_DATA.clone(),
+							expected_results,
+							MESSAGE_COUNT,
+						);
+
+						run_inbound_message_test_suite(suite);
+					}
+				}
+			}
+
+			#[test]
+			fn same_message_multiple_domain_addresses() {
```

Comment:
This one shouldn't pass... it confirms that 2 identical messages from 2 separate domain addresses are not processed separately i.e. there is only 1 inbound entry, for the first one.  We should re-think how we map inbound entries in storage.

Reply:
Good catch!

I think the runtime thought the `RouterProvider` ensures that the case can not exist. From the gateway perspective, you do not have a way to determine it can not exist. But thanks to the `RouterProvider`, if the received messages have different domains, that forcefully implies that they will have different routers. You need to trust that `RouterProvider` is properly mapping domains into routers, though.

So, I would say, from the gateway point of view, it's ok that the same messages from different domains are processed together. That, ideally, would mean that somehow the router handles both domains, so they should be processed together. But really that case will never exists if there are no errors in `RouterProvider`

Reply:
OK, thank you for double-checking and clarifying!